### ISSUE TITLE
Status: code export

### ISSUE THREAD

**Issue by corevo** [SeleniumHQ/selenium-ide#592]
We've just started working on exporting to languages, we will continually update this thread with new information as it becomes available.  

Languages:
- [x] Java (JUnit)
- [x] Python
- [x] C#
- [x] Ruby
- [x] Javascript

We are aware that this feature is very important to the community, but we'd kindly ask to keep all conversation here, until a language is released and then any bug found should raise a ticket.  

We would like to avoid splitting the same conversation into multiple threads.

**Comment by rihp** [comment]
Looking forward to the python version. Thanks for developing this tool

**Comment by DrorCohenCC** [comment]
> Looking forward to the python version. Thanks for developing this tool

likewise! Python or Javascript would be greatly appreciated.

**Comment by kevincvlam** [comment]
+1, can't wait for python. Thanks for all the hard work!

**Comment by jonesp100** [comment]
I'm just about to embark on a large suite of tests - how long do you think the export is going to be?  As I dont want to commit to all these and have to move off to Katalon or some such?  great work as always!!

**Comment by tourdedave** [comment]
@jonesp100 It depends on the language (since we're going in the order listed above, unless others are able to contribute as well), and even then it's hard to know since we're an open source project and timetables & competing priorities can be unpredictable for both contributors and maintainers -- so we can't publicly commit to a timeline.

If your objective is to run your tests cross-browser, in parallel, and/or on CI, then do be sure to check out the Selenium IDE command-line runner while you wait. Details available [here](https://www.seleniumhq.org/selenium-ide/docs/en/introduction/command-line-runner/).

**Comment by jonesp100** [comment]
@tourdedave thanks for the response - appreciate you cant commit but do you have a rough timeline - first language - 1 month, 2 months....?  Thanks!

**Comment by zewa666** [comment]
Hey @corevo, just out of curiosity is it also planned to have a code import sort of thing along the export available. What I mean with that is once you export a code, and do minor modifications according to the style its exported, are you able to re-import it as a side project in Selenium IDE?

Moreover, with regards to the export behavior, do you plan to have a customizable export template per language or is it going to be a fixed set of definitions?

**Comment by BaluErtl** [comment]
What do you recommend, how would be the best to move forward the issue of exporting into PHP, please?

**Comment by tourdedave** [comment]
@jonesp100 ¯\_(ツ)_/¯

@zewa666 No plans for code import at the moment. The initial MVP for code export will just be as-is -- not customizable.

@BaluErtl We're building code export so others can easily contribute other languages. PHP is a tricky one, though, since it's not an officially support language binding. Because of that it's unlikely we'd incorporate it into the mix. But we've discussed the option of making code export for additional languages available through plugins, which could be a nice option to make PHP available if someone wanted to contribute it.

**Comment by zewa666** [comment]
@tourdedave having the code-export configurable via plugins would on the other hand also allow to customize the style of how things get exported. So besides `Java` then there could be an additional plugin to export to `Java using 8 spaces`, don't crucify me for that heretical example ;).

**Comment by AdelSassi** [comment]
Thanx a lot @corevo ! Waiting froward for somme good news about junit export!

**Comment by elinguiuriel** [comment]
> Looking forward to the python version. Thanks for developing this tool

Me to!

**Comment by dstarke1245** [comment]
@tourdedave the .side files aren't affected by this, are they?
So we can still edit our tests there.

and i am also really looking forward to this!  :)

**Comment by corevo** [comment]
Yes, if you keep the `.side` files you can load them in to the IDE change them and re-export them.

**Comment by hectormoreno87** [comment]
Great , waiting for C# exporting. 

Thank you all.

**Comment by InassDrissi** [comment]
waiting for java exporting , thank you !

**Comment by steffenkj** [comment]
waiting for python export :)

**Comment by sah-anshu** [comment]
waiting for python export :) too.

**Comment by vvccool** [comment]
等待python导出

**Comment by Sartek158** [comment]
Export to Java is great!

**Comment by stebli** [comment]
Why not allow users to specify their own export actions? Instead of hardcoding this in js files rather allow a config which can be modified to cater for different permutations of various frameworks. 

so for a click i can opt to use a specific bit of code, specified in a text file for the click, instead of all these repetitive lines being exported. Just an idea... would then solve all "languages" issues as this would be abstracted and we can key/value pair the stuff.

**Comment by sah-anshu** [comment]
fyi- Just for confirmation, is export to Java Working?

**Comment by tourdedave** [comment]
@sah-anshu Yes, export is available for Java JUnit.

@stebli Code export is doing a lot more than just exporting individual commands. Allowing user customizations through a separate file isn't enough to solve for all languages -- especially if it puts the burden onto the user. But it's an interesting idea - thanks!

**Comment by MisinformedDNA** [comment]
@tourdedave @stebli I think this would be best solved by good guidance on how to create plugins to export code in different languages/test frameworks.

**Comment by bcmoney** [comment]
@tourdedave Is this being added to the Browser plugin/extension in the next release as well, or only via passing those parameters to the CLI tool?

**Comment by stebli** [comment]
> @tourdedave @stebli I think this would be best solved by good guidance on how to create plugins to export code in different languages/test frameworks.

is this an option right now? Will it be limited to JS?

**Comment by corevo** [comment]
The plan is to allow 3rd party plugins to export code, we will be exporting to the official bindings, this way you could write a plugin to export to webdriver.io or such.

**Comment by stebli** [comment]
> The plan is to allow 3rd party plugins to export code, we will be exporting to the official bindings, this way you could write a plugin to export to webdriver.io or such.

awesome. await that functionality....veery useful

**Comment by upenk910** [comment]
Do anyone know when exporting to java will be fixed on the new selenium IDE

**Comment by sah-anshu** [comment]
Eagerly Waiting for Python Code Export.

**Comment by tourdedave** [comment]
@upenk910 What do you mean? Java export is available in the latest version (3.6.0).

**Comment by upenk910** [comment]
i don't see the option can you tell me where is option to export in java
the test case

**Comment by rori4** [comment]
just right click on your test and there will be an export option. I have just exported one example test to java so I can confirm it is working 
![image](https://user-images.githubusercontent.com/22402693/57019320-cae7e000-6c2e-11e9-9235-23c969c15cd7.png)

**Comment by upenk910** [comment]
Thanks, since to GUI of the selenium IDE has changed from the previous
versions, i couldn't find it.

**Comment by tourdedave** [comment]
The docs have been updated to cover code export and include instructions for contribution a new language/framework.

https://www.seleniumhq.org/selenium-ide/docs/en/introduction/code-export/

**Comment by stebli** [comment]
> The docs have been updated to cover code export and include instructions for contribution a new language/framework.
> 
> https://www.seleniumhq.org/selenium-ide/docs/en/introduction/code-export/

Hi tourdedave,

I followed these instructions but do not see the new item in the export list. Is this a complete guide? 
I see issues (i.e. missing aliases under code-export\node_modules\xxx) that arent mentioned in the guide if i just look at the folder structures.

**Comment by tourdedave** [comment]
To have it show up in the list you need to be sure to complete [step 5](https://www.seleniumhq.org/selenium-ide/docs/en/introduction/code-export/#5-add-it-to-the-mix). If you're getting build warnings then try `cd`ing into your package and running `yarn build`.

**Comment by stebli** [comment]
> step 5

have done that yes...

yarn build yields 

yarn build
yarn run v1.16.0
$ babel -d dist src --ignore=__test__
src/command.js -> dist/command.js
src/hook.js -> dist/hook.js
src/index.js -> dist/index.js
src/location.js -> dist/location.js
src/selection.js -> dist/selection.js
✨  Done in 0.92s.

when running yarn build:ext:prod from root i do see these warnings, could they be the issue?

WARNING in ./neo/stores/seed.js
17:496-509 "export 'CommandsArray' was not found in '../models/Command'
 @ ./neo/stores/seed.js
 @ ./neo/containers/Panel/index.jsx
 @ ./neo/containers/Root/index.jsx
 @ multi react-hot-loader/patch ./neo/containers/Root

WARNING in ./neo/stores/seed.js
17:535-548 "export 'CommandsArray' was not found in '../models/Command'
 @ ./neo/stores/seed.js
 @ ./neo/containers/Panel/index.jsx
 @ ./neo/containers/Root/index.jsx
 @ multi react-hot-loader/patch ./neo/containers/Root

**Comment by tourdedave** [comment]
No, those warnings are OK. Can you publish your fork and send a link?

**Comment by stebli** [comment]
Unfortunately not(did not fork) , company is locked down and we cant remote/upstream to github.

**Comment by tourdedave** [comment]
Can you create a gist with the contents of your `selenium-ide/packages/code-export/src/index.js`? The code export UI is built using the `availableLanguages` object in that file. That's the first place I'd look.

**Comment by stebli** [comment]
https://gist.github.com/stebli/81ef417e2ad4d41cf9f961cbdb52dab3

**Comment by tourdedave** [comment]
@stebli That looks correct. Make sure that 1) you've added your package as a dependency in `code-export` (e.g., [here](https://github.com/SeleniumHQ/selenium-ide/blob/c55c556ffc947fd3f6ee8ab317915c6f879a88dc/packages/code-export/package.json#L22)) and 2) you're able to run `yarn` from the root of the project before building the extension (e.g., `yarn build:ext:prod` or `yarn watch`).

I've updated step 1 to include these bits, which were omitted.

**Comment by tourdedave** [comment]
@stebli Can you open a new issue? It will be better to track the progress there.

**Comment by stebli** [comment]
@tourdedave https://github.com/SeleniumHQ/selenium-ide/issues/689 created.

**Comment by tourdedave** [comment]
\o/ Thanks!

**Comment by tourdedave** [comment]
I just submitted a PR for Python support (e.g., #719).

It will be available in the next release.

<img width="662" alt="Screen Shot 2019-06-13 at 15 51 35" src="https://user-images.githubusercontent.com/457137/59433927-29d67280-8df3-11e9-9745-dd023e2b1df7.png">

**Comment by tourdedave** [comment]
FYI - code export for Python was released yesterday.

**Comment by e-omahony** [comment]
@tourdedave, the python export is awesome, thanks to you and everyone else involved!

**Comment by ctrngk** [comment]
Yes we have python code, but I simply do not know how to run the script. There is only a class implementation there

**Comment by CEN888** [comment]
Thanks for your work @tourdedave it's been a great help .

**Comment by darkartswizard** [comment]
I saw it yesterday!
This is awesomeness!

**Comment by darkartswizard** [comment]
Actually I just tried it on Chrome and Firefox and can’t get export to work
for Python or Java.

**Comment by tourdedave** [comment]
@darkartswizard, et al. If you're having a problem with code export, please open an issue and follow the template ([link](https://github.com/SeleniumHQ/selenium-ide/issues/new?template=bug.md)).

**Comment by tourdedave** [comment]
Support for JavaScript with Mocha has been released (in version `3.10.0`).

**Comment by darkartswizard** [comment]
My bad. I thought this was in response  of the Pull Request. Will do.

**Comment by darkartswizard** [comment]
#730 posted

**Comment by catalin-florescu** [comment]
When is gonna be fixed?

**Comment by tourdedave** [comment]
@catalin-florescu What do you mean? If you're referring to #730, it's fixed. If you are experiencing an issue with code export, please open a new issue and fill out the issue template ([link](https://github.com/SeleniumHQ/selenium-ide/issues/new?template=bug.md)).

**Comment by catalin-florescu** [comment]
@tourdedave i filled [here](https://github.com/SeleniumHQ/selenium-ide/issues/742) the issue. Thanks.

**Comment by hectormoreno87** [comment]
> Great , waiting for C# exporting.
> 
> Thank you all.

Still waiting ;(

**Comment by fooschamp** [comment]
Looking forward to C# export.  Thanks for all your hard work!

**Comment by tourdedave** [comment]
Some related work as part of code export that's language agnostic: #767 and #742.

**Comment by tourdedave** [comment]
Fixes for #767 and #742 have been released in v3.12.2 of the IDE.

**Comment by seanpoulter** [comment]
Have you considered support for the Robot Framework Selenium library @corevo and @tourdedave (or _would_ you consider it)?  Would you welcome a PR that adds support? It seems like you've laid a solid foundation with @tourdedave's `junit`, `mocha`, and `pytest` PRs so we (the community) could build out a language port in the same beautiful TDD style. What do you think?

For context, I'm at my second workplace where folks without a lot of programming experience are being eased into test automation using Robot Framework's keyword-driven approach. They're using the Selenium IDE to get familiar with Selenium and selectors. It's working well so far 👏 .

**Comment by u1266221** [comment]
When are you planning to deploy C# code exporting? looking forward to this :)

**Comment by georgemjohnson11** [comment]
I'm not sure for an ETA. I went with an xUnit approach that I'm still resolving some small kinks I'd like to set before merging this in since xUnit doesn't have methods called for startup and completion. This is currently building and testing fine locally, but the build server is acting up a bit.

**Comment by tourdedave** [comment]
I've just merged code export for C# with NUnit. It will go out in the next release later this week. There's still some work to land @georgemjohnson11's PR for C# with XUnit, which will end up sharing some code between both packages. It will be nice to see since it will set a good precedent for how we will handle multiple frameworks per programming language going forward.

**Comment by georgemjohnson11** [comment]
@tourdedave  What work is there? It's been sitting there for 3 days without a comment on the most recent commits

**Comment by tourdedave** [comment]
We’ll want to share a lot of the same emitted code between the NUnit and
XUnit packages (e.g., commands, locator strategies, etc). This will likely
be done through a csharp-commons package which will require a decent amount
of refactoring.

**Comment by georgemjohnson11** [comment]
It would be nice to have a commons package, but I'm not going to do it. I'll let the community pick it up if they want but that's not the issue I'm solving for.

This is not what the instructions (that are 4 months old and you wrote) detail: https://www.seleniumhq.org/selenium-ide/docs/en/introduction/code-export/ https://github.com/SeleniumHQ/selenium-ide/commits/v3/docs/introduction/code-export.md) Please update these so other potential contributors don't waste their time trying to implement a testing framework. 

I'll find another open source project where people don't ignore pull requests so they can rush to contribute bugs like this: https://github.com/SeleniumHQ/selenium-ide/commit/1bfcd258d4af6b34a4536e8a642fdfb347ddf13a Thanks @tourdedave

**Comment by tourdedave** [comment]
You misunderstand, @georgemjohnson11. We appreciate your PR and intend to
merge it. The work I described is something I am planning to do but will
take some time. For context: the work to implement C# NUnit started last
month but was stalled while I was out on paternity leave and is my primary
focus at the moment. The commit you referenced is not a bug. It was a
necessary change to update my branch to work with the latest version of
side-utils which was updated while I was away. As architectural changes
start to shake out to support multiple frameworks per language then the
docs will be updated. But as for now, the docs are current and you didn’t
waste your time - we’re blazing a trail and I think that’s something to be
excited about. You did something I hope more people will do in the future
and I’m sure the community will appreciate it. It will just take some time
to make the necessary changes and test everything.

**Comment by tourdedave** [comment]
I was mistaken about the upcoming release. @corevo reminded me that we're waiting until November 7th to do the next release. I'll start working on the refactoring to support both C# framework in the meantime.

**Comment by tourdedave** [comment]
C# NUnit & C# xUnit are done and merged. They'll go out in the next release.

<img width="952" alt="Screen Shot 2019-10-24 at 9 17 29" src="https://user-images.githubusercontent.com/457137/67458547-1fe53b80-f63f-11e9-9bf3-998e23ecff58.png">

**Comment by tourdedave** [comment]
I'm also working on export for Ruby RSpec, which should go out in the release too.

**Comment by tourdedave** [comment]
<img width="662" alt="Screen Shot 2019-11-10 at 11 58 12" src="https://user-images.githubusercontent.com/457137/68542114-65c93000-03b1-11ea-923c-1620fb2ee406.png">

The remaining languages have just been released in `v3.15.0` (available now from both the Chrome and Firefox web stores).

Each of the officially supported languages for Selenium are now supported (for at least one test framework) in code export. If you have a problem with code export or have an idea for how to improve it -- please open an issue and follow the provided template ([link](https://github.com/SeleniumHQ/selenium-ide/issues/new/choose)).

**Comment by darkartswizard** [comment]
Sweet!!!

**Comment by DGrantham0283** [comment]
This is really great - Im very happy about the C# export

**Comment by Nabeelhassan** [comment]
Please look into this #880. Thank you.

**Comment by defro** [comment]
Hello @tourdedave 

You wrote 
> @BaluErtl We're building code export so others can easily contribute other languages. PHP is a tricky one, though, since it's not an officially support language binding. Because of that it's unlikely we'd incorporate it into the mix. But we've discussed the option of making code export for additional languages available through plugins, which could be a nice option to make PHP available if someone wanted to contribute it.

I'm very insterested to contribute to this project and create a export for PHP [Codeception](https://codeception.com/) for Selenium IDE.
So what did you decide : a plugin or incorporation ?
